File: 2020/18_OperationOrder/homework.ts
// ======================================================================
// Operation Order
//   Advent of Code 2020 Day 18 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           h o m e w o r k . t s
//
// A solver for the Advent of Code 2020 Day 18 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------
import { Expression } from './expression';

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------

// ======================================================================
//                                                               Homework
// ======================================================================

export class Homework {
  // Object for Operation Order
  text: string[];

  part2: boolean;

  expressions: Expression[];

  constructor(text: string[], part2 = false) {
    // Create a Homework object

    // 1. Set the initial values
    this.text = text === undefined ? [] : text;
    this.part2 = part2 === undefined ? false : part2;
    this.expressions = [];

    // 2. Process text (if any)
    if (this.text.length !== 0) {
      for (let indx = 0; indx < this.text.length; indx += 1) {
        this.expressions.push(new Expression(this.text[indx], this.part2));
      }
    }
  }

  evaluateAll(): number {
    // Returns the sum of evaluatin all of the expressions
    // 1. Start with nothing
    let result = 0;

    // 2. Loop for all the expressions
    for (let indx = 0; indx < this.expressions.length; indx += 1) {
      const expression = this.expressions[indx];

      // 3. Evaluate the expression
      const value = expression.evaluate();

      // 4. Accumulate the value of the expressions
      result += value;
    }

    // 5. Return the sum of the evaluated expressions
    return result;
  }

  solution(verbose = false, limit = 0): number {
    if (verbose) console.log(`solution: ${limit}`);
    return this.evaluateAll();
  }

  partOne(verbose = false, limit = 0): number {
    // Returns the solution for part one

    return this.solution(verbose, limit);
  }

  partTwo(verbose = false, limit = 0): number {
    // Returns the solution for part two

    // 1. Return the solution for part two
    return this.solution(verbose, limit);
  }
}

// ======================================================================
// end                      h o m e w o r k . t s                     end
// ======================================================================

File: 2021/18_Snailfish/homework.py
# ======================================================================
# Snailfish
#   Advent of Code 2021 Day 18 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         h o m e w o r k . p y
# ======================================================================
"A solver for the Advent of Code 2021 Day 18 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import number

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

# ======================================================================
#                                                               Homework
# ======================================================================


class Homework(object):   # pylint: disable=R0902, R0205
    "Object for Snailfish"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.pairs = []
        self.sumation = []

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            for line in text:
                self.pairs.append(number.Number(text=line, part2=part2))

    def sum_them_up(self):
        "Sum the homework numbers"

        # 1. Start with the first one
        result = number.Number(text=str(self.pairs[0]), part2=self.part2)

        # 2. Loop for the rest of the numbers
        for pair in self.pairs[1:]:

            # 3. Add it in
            result.add(pair.pair)

        # 4. Save the sumation
        self.sumation = result

    def largest_magnitude(self):
        "Determine the largest magnitude from adding any two numbers"

        # 1. Start with nothing
        result = 0

        # 2. Loop for all of the pairs of pairs
        for pair1 in self.pairs:
            for pair2 in self.pairs:
                if pair1 == pair2:
                    continue

                # 2. Add pair2 to pair1 and check the magnitude
                test = number.Number(text=str(pair1))
                test.add(pair2.pair)
                mag = test.magnitude()
                if mag > result:
                    result = mag

                # 3. Add pair1 to pair2 and check the magnitude
                test = number.Number(text=str(pair2))
                test.add(pair1.pair)
                mag = test.magnitude()
                if mag > result:
                    result = mag

        # 4. Return the largest magnitude
        return result

    def part_one(self, verbose=False, limit=0):
        "Returns the solution for part one"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part one
        self.sum_them_up()
        return self.sumation.magnitude()

    def part_two(self, verbose=False, limit=0):
        "Returns the solution for part two"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part two
        return self.largest_magnitude()


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                      h o m e w o r k . p y                     end
# ======================================================================
